scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -1038,6 +1041,87 @@
                               :number-unmatched (- (count offers) (count offers-scheduled))
                               :stats (offers->stats offers)}})
 
+          ; We want to log warnings when jobs have gone unmatched for a long time.
+          ; In order to do this, we keep track, per pool, of the jobs that did not
+          ; get matched to an offer, along with how many matching cycles they've
+          ; gone unmatched for. The amount of data we store is relatively small;
+          ; it's O(# pools * # unmatched considerable jobs). If a job uuid does
+          ; get matched, we stop storing it. We never store job uuids that were
+          ; not considerable in the first place.
+          (let [unmatched-job-uuids
+                (set/difference
+                  (->> considerable-jobs (map :job/uuid) set)
+                  (set matched-job-uuids))
+                ; There are two configuration knobs we can tweak:
+                ; - unmatched-cycles-warn-threshold:
+                ;   the # of consecutive unmatched matching cycles we care about
+                ; - unmatched-fraction-warn-threshold:
+                ;   the fraction of considerable jobs that have gone unmatched for
+                ;   at least unmatched-cycles-warn-threshold beyond which we will
+                ;   warn
+                {:keys [unmatched-cycles-warn-threshold
+                        unmatched-fraction-warn-threshold]}
+                (config/offer-matching)]
+            (swap!
+              ; This atom's value is a map of the following shape:
+              ;
+              ; {"pool-1" {job-uuid-a count-a
+              ;            job-uuid-b count-b
+              ;            ...}
+              ;  "pool-2" {job-uuid-c count-c
+              ;            job-uuid-d count-d
+              ;            ...}
+              ; ...}
+              ;
+              ; where the counts are the numbers of consecutive
+              ; matching cycles that the job has gone unmatched
+              pool-name->unmatched-job-uuid->unmatched-cycles-atom
+              (fn [m]
+                (let [unmatched-job-uuid->unmatched-cycles
+                      (pc/map-from-keys
```

Comment:
Would like some text in here explaining no leak.

"""This code doesn't leak jobs and grow forever. This builds a new map from scratch of size at most unmatched-job-uuids which is less than num-considerable, which goes in. That new map gets assoc'ed in replacing the existing job-uuid->unmatched-cycles submap in pool-name->unmatched-job-uuid->unmatched-cycles-atom. Which means we won't leak historic jobs."""

Reply:
Good call, done